File: boj/1759.암호만들기/wg.py
from itertools import *
import sys
input = sys.stdin.readline


L, C = map(int, input().split())
alphabets = list(input().split())

tuple_pw_list = list(combinations(alphabets, L))

pw_list = []
vowel = ['a', 'e', 'i', 'o', 'u']

for elem in tuple_pw_list:
  elem = list(elem)
  vowel_cnt = 0
  consonant_cnt = 0

  for char in elem:
    if char in vowel:
      vowel_cnt += 1
    else:
      consonant_cnt += 1
  
  if vowel_cnt >= 1 and consonant_cnt >= 2:
    pw_list.append(''.join(sorted(elem)))


pw_list.sort()

for elem in pw_list:
  print(elem)

File: boj/1759.암호만들기/wg.js
const fs = require('fs');
// const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const input = fs.readFileSync('input.txt').toString().trim().split('\n');

const [L, C] = input[0].split(' ').map(Number);
const alphabets = input[1].split(' ').sort();

const vowels = ['a', 'e', 'i', 'o', 'u']

// 백트래킹?
let result = []
let stack = []

const dfs = (start, vowelCnt, consonantCnt) => {
  if (stack.length == L) {
    if (vowelCnt >= 1 && consonantCnt >= 2) {
      result.push(stack.join(''))
    }
    return;
  }

  const need = L - stack.length

  for (let i = start; i <= alphabets.length - need; i++) {
    let char = alphabets[i]

    stack.push(char)

    if (vowels.includes(char)) {
      dfs(i + 1, vowelCnt + 1, consonantCnt)
    } else {
      dfs(i + 1, vowelCnt, consonantCnt + 1)
    }

    stack.pop()
  }
}


dfs(0, 0, 0)

console.log(result.join('\n'))
